### PR TITLE
Fix: interruption signal not terminating the process when spamming `<C-c>`

### LINKED ISSUE #477: [Bug]: inconsistent SIGINT behavior
### Contact Details

teocns@gmail.com

### 📦 Package Version

latest

### 🎞️ Framework Version

N/A

### 🔎 Describe the Bug

```
import os
import time

import agentops
import dotenv

dotenv.load_dotenv()

agentops.init(api_key=os.environ["AGENTOPS_API_KEY"])


while True:
    time.sleep(0.2)
```

1. Run the script

2. Spam `<C-c>`

3. Won't [always] terminate

```
🖇 AgentOps: Session Replay: https://app.agentops.ai/drilldown?session_id=6f8b6bee-33c5-4288-9377-c1f2bb0d264f
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
^C🖇 AgentOps: SIGINT detected. Ending session...
... (Goes on forever)
```

### 🤝 Contribution

- [X] Yes, I'd be happy to submit a pull request with these changes.
- [ ] I need some guidance on how to contribute.
- [ ] I'd prefer the AgentOps team to handle this update.

### PR DESCRIPTION
**Pull request by teocns**
Fixes #477

### CODE REVIEW

File: agentops/client.py [hunk 1]
```diff
@@ -318,6 +318,8 @@ def cleanup(end_state: str = "Fail", end_state_reason: Optional[str] = None):
                         end_state_reason=end_state_reason,
                     )
 
+        __handled = False
```
Comment (areibman) [inline agentops/client.py]: Minor nit: rename to `__signal_handled`

**Comment by teocns** [comment]
After further consideration, addressing this at the root is essential. While this fix works, it only addresses the surface issue.

Looking at the [Session](/AgentOps-AI/agentops/blob/main/agentops/session.py) implementation everything made more sense; `threading` components are being used, indicating that somewhere the Session state is left in an inconsistent state, specifically and presumably happening as Control+C will initiate re-entry into a circuit whose state was left "open"; this is very common side effects caused by improper `threading.Lock` implementation, like in this case.

**Comment by areibman** [comment]
Cool fix. I've had this issue a number of times but hard to replicate and pin down when/why. A test script that's able to replicate + show the fix would be much appreciated

**Comment by teocns** [comment]
Closing in favor of #486

**Comment by teocns** [comment]
Let's merge this for now until core refactoring is ready

**Review by areibman — APPROVED** [review]
Merge for now

**Comment by teocns** [comment]
Not convincing, need to review

**Comment by teocns** [comment]
Fixed in #529

**Pull request closed without merging**